release-19.2: sql: move notifying StatsRefresh of new table to post-commit hook

ajwerner opened the pull request:
Backport 1/1 commits from #47718.

/cc @cockroachdb/release

---

Prior to this change, the StatsRefresher was being notified of a new table
during the execution of the createTable node, before the creating transaction
had committed. Prior to #46170, the StatsRefresher was likely to block on the
intent of the creating transaction. After that change, the StatsRefresher might
not discover the new table because the creating transaction is much more likely
to get pushed further into the future, past the read of the stats refresher.

Release note (bug fix): Fix rare bug where stats were not automatically
generated for a new table.

Commit:
sql: move notifying StatsRefresh of new table to post-commit hook

Prior to this change, the StatsRefresher was being notified of a new table
during the execution of the createTable node, before the creating transaction
had committed. Prior to #46170, the StatsRefresher was likely to block on the
intent of the creating transaction. After that change, the StatsRefresher might
not discover the new table because the creating transaction is much more likely
to get pushed further into the future, past the read of the stats refresher.

Release note (bug fix): Fix rare bug where stats were not automatically
generated for a new table.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48028)
<!-- Reviewable:end -->

Commit:
importccl: move stats refresh to OnTerminal hook

In 19.2 and prior, the jobs state machine included an OnTerminal hook which was
called after the transaction which marked the job as succeeded commits. In that
older version, the OnSuccess hook took a transaction and in that transaction
published new table versions. This was problematic because the stats refresher
may not see the newly published table versions. While the OnTerminal hook is
not guaranteed to run, it generally does and we don't really have a way to
guarantee that such a hook runs in 19.2.

Release note (bug fix): Trigger stats refreshing after completion of an import
job after the transaction finishing the job commits, fixing a rare bug where
stats refreshes could fail.

ajwerner commented:
@andreimatei I added a second commit to this backport that only applies to 19.2 to fix https://github.com/cockroachdb/cockroach/issues/43555#issuecomment-622555597

andreimatei commented:
LGTM

ajwerner commented:
`TestCreateStatsAfterImport` failed under stress on the 19.2 release branch before adding that commit and passes with it.
